Cargo support

boxdot opened the pull request:
For details cf. the commit message.

Resolves #724

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1602)
<!-- Reviewable:end -->

sschuberth commented:
Thanks @boxdot! Would you mind squashing all your WIP commits in the PR, so the PR consists of only a single commit to start with, and use the text from the PR as the commit message of that commit? We try to capture all valuable information in Git, and not leave important bits behind in the GitHub conversation UI.

boxdot commented:
@sschuberth: Squashed with embedded description. Updated the description of the PR.

sschuberth commented on .appveyor.yml:
```diff
@@ -56,6 +57,10 @@ install:
   - ps: Start-FileDownload "https://dl.google.com/android/repository/sdk-tools-windows-$env:ANDROID_SDK_VERSION.zip"
   - 7z x sdk-tools-windows-%ANDROID_SDK_VERSION%.zip -o%ANDROID_HOME% > nul
   - yes | %ANDROID_HOME%\tools\bin\sdkmanager.bat platform-tools
+  # Install rust
```
There are Rust packages for Chocolatey (for both MSVC or GNU ABI), can we use these instead? See https://chocolatey.org/packages?q=rust.

boxdot replied:
It might work however I did not see a single usage of `choco` in you appveyor script. I used the snippet from the official rustup app veyor configuration. Should I still switch to `choco`?

sschuberth replied:
See the calls to `cinst`, which is the short form of `choco install` :wink: If that works I'd prefer that to reduce the clutter a bit.

sschuberth commented on .travis.yml:
```diff
@@ -53,6 +54,8 @@ install:
   - unzip -q sdk-tools-linux-$ANDROID_SDK_VERSION.zip -d $ANDROID_HOME
   - export SDKMANAGER_OPTS="--add-modules java.xml.bind"
   - yes | $ANDROID_HOME/tools/bin/sdkmanager --verbose "platform-tools"
+  - curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --default-toolchain $RUST_VERSION
```
This line looks more complex than the one at https://forge.rust-lang.org/other-installation-methods.html, why?

boxdot replied:
This the official line from https://rustup.rs/ which is more secure, since it allows only tlsv1.2 https connections.

sschuberth commented on analyzer/src/funTest/assets/projects/synthetic/cargo-expected-output.yml:
```diff
@@ -0,0 +1,95 @@
+    description: "Synchronization primitives based on spinning.\nThey may contain\
+      \ data, are usable without `std`,\nand static initializers are available.\n"
+    homepage_url: ""
+    binary_artifact:
```
Why's there no binary artifact? Does Rust always build from source?

boxdot replied:
Cargo does not support distribution of binary artifacts. I am not sure if there is anything planned for the future. At least I could not find any related issues.

sschuberth commented on analyzer/src/funTest/assets/projects/synthetic/cargo/.gitignore:
```diff
@@ -0,0 +1 @@
+target
```
Nit: Missing final newline.

sschuberth commented on analyzer/src/funTest/kotlin/CargoSubcrateTest.kt:
```diff
@@ -0,0 +1,111 @@
+class CargoSubcrateTest : StringSpec() {
+    private val projectDir = File(
+        "src/funTest/assets/projects/synthetic/cargo-subcrate"
+    ).absoluteFile
```
Nit: This looks like it should fit into one line.

boxdot replied:
I used the official intellij formatter, so I guess either I had a wrong configuration or there is none.

boxdot replied:
Never mind, you said below that it is 120 chars.

sschuberth commented on analyzer/src/funTest/kotlin/CargoSubcrateTest.kt:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    init {
```
I wasn't aware it even works to have multiple `init` blocks 😄 But for consistency we should merge these into only a single `init` block.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+import com.here.ort.utils.textValueOrEmpty
+
+import com.vdurmont.semver4j.Requirement
+import com.moandjiezana.toml.Toml
```
Nit: We separate `com` imports from different groups by a blank line, so please add a blank line before this one.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) =
+            Cargo(managerName, analyzerRoot, analyzerConfig, repoConfig)
```
Nit: We changed that formatting recently, please put this into the line above.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+
+    override fun getVersionRequirement(): Requirement = Requirement.buildStrict(REQUIRED_CARGO_VERSION)
+
+    private fun runMetadata(workingDir: File): String {
```
Nit: Could be an expression function (using `=` to define the function body).

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        return run(workingDir, "metadata", "--format-version=1").stdout
+    }
+
+    private fun extractCargoId(node: JsonNode): String {
```
Nit: Could be an expression function.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        return node["id"].textValue()!!
+    }
+
+    private fun extractPackageId(node: JsonNode) = Identifier(
```
Nit: Formatting should be like

    private fun extractPackageId(node: JsonNode) =
        Identifier(
            // ...
        )

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+
+    private fun extractVcsInfo(node: JsonNode): VcsInfo {
+        val url = extractRepositoryUrl(node)
+        val type = if (url != null) "git" else "" // for now cargo supports only git
```
More Kotlinish: `url?.let { "git" } ?: ""`

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        return VcsInfo(type, url.orEmpty(), revision = "")
+    }
+
+    private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
```
Nit: Could be an expression function.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
+        return node["license"].textValueOrEmpty().split("/")
+            .map { it.trim() }
+            .filter { it.isNotBlank() }
```
As you already trim before, you can say `.filter { it.isNotEmpty() }` or `.filterNot { it.isEmpty() }` here.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    ): RemoteArtifact? {
+        if (node["source"].textValueOrEmpty()
+            == "registry+https://github.com/rust-lang/crates.io-index"
+        ) {
```
That should fit into a single line (we limit to 120 chars per line).

sschuberth replied:
To reduce nesting, how about returning `null` early here if `!=`?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        if (node["source"].textValueOrEmpty()
+            == "registry+https://github.com/rust-lang/crates.io-index"
+        ) {
+            val name = node["name"]?.asText() ?: return null
```
For nodes known to be text nodes, we should prefer to use `textValue()` instead of `asText()`.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            sourceArtifact = extractSourceArtifact(node, hashes) ?: RemoteArtifact.EMPTY,
+            homepageUrl = "",
+            vcs = vcsInfo,
+            vcsProcessed = vcsInfo.normalize()
```
That's the default value of `vcsProcessed` and thus can be omitted. Consequently, the `vcsInfo` variable can be inlined, and this function be made an expression function.

boxdot replied:
Ah, nice!

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    private fun checksumKeyOf(metadata: JsonNode): String {
+        val id = metadata["id"]!!.textValue()
```
Maybe `textValueOrEmpty()` instead of the unchecked `!!` cast?

sschuberth replied:
Was this supposed to be changes? If not, why not?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            }
+            workingDir = workingDir.parentFile
+        }
+        // we reached the root directory we are analyzing
```
Nits: Blank line before this one, start comment with upper case letter and end with a dot.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        var workingDir = definitionFile.parentFile
+        while (workingDir != analyzerRoot.parentFile) {
+            val lockfile = File(workingDir, "Cargo.lock")
+            if (lockfile.exists()) {
```
Better use `lockfile.isFile` to not only check whether is exists, but also whether it is a file (and not e.g. a directory).

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            workingDir = workingDir.parentFile
+        }
+        // we reached the root directory we are analyzing
+        throw IllegalArgumentException("missing Cargo.lock file")
```
Nit: "Missing" (start with upper case).

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        val contents = Toml().read(lockfile)
+        val metadata = contents.getTable("metadata") ?: return emptyMap()
+        val metadataMap = metadata.toMap()
+        return metadataMap.asSequence().mapNotNull {
```
The view as a sequence doe snot serve any purpose here, or? How about this instead:

        val metadataMapNotNull = mutableMapOf<String, String>()

        metadataMap.forEach { (key, value) ->
            if (key != null && (value as? String) != null) {
                metadataMapNotNull[key] = value
            }
        }

        return metadataMapNotNull

boxdot replied:
not very functional, but ok

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+
+        return resolveDependenciesOf(
+            root, nodes, packages, filter
+        ).dependencies
```
Nit: Could write this as a one-liner.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        packages: Map<String, Package>,
+        filter: (pkgId: String, depId: String) -> Boolean
+    ): PackageReference {
+        val node = nodes.find { it["id"].textValueOrEmpty() == id }!!
```
If you're sure there's exactly one match, use `.single {}` instead of `.find {}!!`.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        filter: (pkgId: String, depId: String) -> Boolean
+    ): PackageReference {
+        val node = nodes.find { it["id"].textValueOrEmpty() == id }!!
+        val depsReferences = node["dependencies"].asSequence()
```
As you're always iterating over all items, is there a benefit of turning this into a sequence here?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            .map { it.textValue()!! }
+            .filter { filter(id, it) }
+            .map { resolveDependenciesOf(it, nodes, packages, filter) }
+        val pkg = packages[id]!!
```
Use `getValue()` instead of cast.

sschuberth replied:
This actually is also hinted by IDE inspection and should be changed.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    // We treat all path dependencies inside of the analyzer root as project dependencies
+    private fun isProjectDependency(id: String): Boolean {
+        val pathRegex = Regex("""^.*\(path\+file://(.*)\)$""")
+        val match = pathRegex.matchEntire(id)?.groups?.get(1)
```
More Kotlin-ish:

        pathRegex.matchEntire(id)?.groups?.get(1)?.let { match ->
            val packageDir = File(match.value)
            return packageDir.startsWith(analyzerRoot)
        }

sschuberth replied:
Ping?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    //
+    // We treat all path dependencies inside of the analyzer root as project dependencies
+    private fun isProjectDependency(id: String): Boolean {
+        val pathRegex = Regex("""^.*\(path\+file://(.*)\)$""")
```
This could go to some private constant so the Regex is not recompiled on each function call.

sschuberth replied:
Ping?

boxdot replied:
Sorry, somehow I missed several comments.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            val packageDir = File(match.value)
+            return packageDir.startsWith(analyzerRoot)
+        }
+        return false
```
Nit: Blank line before this one.

sschuberth replied:
Ping?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+    override fun resolveDependencies(definitionFile: File): ProjectAnalyzerResult? {
+        log.info { "Resolving dependencies for: '$definitionFile'" }
+
+        // get project name; if none => we have a workspace definition => return null
```
Nit: Comments should start with a capital letter and end with a dot, just like written sentences.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        val hashes = readHashes(findCargoLock(definitionFile))
+
+        // collect all packages
+        val packages: Map<String, Package> = metadata["packages"].asSequence().associateBy(
```
Nit: Could drop the explicit type declaration.

sschuberth replied:
Ping?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        // collect all packages
+        val packages: Map<String, Package> = metadata["packages"].asSequence().associateBy(
+            { extractCargoId(it) },
+            { extractPackage(it, hashes) })
```
Nit: The closing parenthesis should go to the next line.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+            { extractPackage(it, hashes) })
+
+        // resolve dependencies tree
+        val dependencies = resolveDependenciesTree(metadata, packages,
```
Formatting:

        val dependencies = resolveDependenciesTree(metadata, packages) { id, devId ->
            !isDevDependencyOf(id, devId, metadata)
        }

sschuberth replied:
Similar below.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+
+        // resolve project
+        val projectPkg = packages.values.asSequence().find {
+            val pkgId = it.id
```
Instead of this, use `{ pkg ->` above and then `pkg.id` below.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,290 @@
+        )
+
+        val nonProjectPackages = packages
+            .filter { !isProjectDependency(it.key) }
```
Nit: Maybe `.filterNot { isProjectDependency(it.key) }` is a tad nicer.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -141,7 +141,12 @@ class Downloader {
         // Try downloading from VCS.
         try {
-            return downloadFromVcs(target, targetDir, allowMovingRevisions)
+            // Cargo in general builds from source tarballs, so we prefer source artifacts to VCS.
```
I've not yet reviewed changes in this file, will do that later.

sschuberth commented:
While I had a bunch of comments, all in all very nice work @boxdot!

boxdot commented:
Thanks for the review! All resolved comments are applied.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,280 @@
+        if (node["source"].textValueOrEmpty() != "registry+https://github.com/rust-lang/crates.io-index") {
+            return null
+        }
+        val name = node["name"]?.textValue() ?: return null
```
Nit: Blank line before this one. (We generally use a blank line to separate a following statement from a preceding closing block.)

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,280 @@
+    ): PackageReference {
+        val node = nodes.single { it["id"].textValueOrEmpty() == id }
+        val depsReferences = node["dependencies"]
+            .map { it.textValue()!! }
```
Again, maybe `textValueOrEmpty()`?

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,280 @@
+
+    private fun isDevDependencyOf(id: String, depId: String, metadata: JsonNode): Boolean {
+        val packages = metadata["packages"]
+        val pkg = packages.asSequence().find { it["id"].textValueOrEmpty() == id }!!
```
Forgot to also use `single {}` here and below?

sschuberth commented:
Second round of review done. I expected a third round as I still need to review the changes in the last file. Please simply squash all review-related changes into the commit where the reviewed code was introduced, i.e. in this case your first commit.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -335,6 +340,12 @@ class Downloader {
                     }
                 }
+            } else if (target.id.type == "Cargo") {
+                // without following redirect Cargo's package do not have an extension
```
Which redirect? HTTP? Should we instead configure our HTTP client to follow redirects?

boxdot replied:
I solved this by using for the name of the archive the last bit of the url which we get from the last redirect. This block is gone now.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -335,6 +340,12 @@ class Downloader {
+            } else if (target.id.type == "Cargo") {
+                // without following redirect Cargo's package do not have an extension
+                val tarFile = File(
+                    sourceArchive.parentFile, sourceArchive.nameWithoutExtension + ".tgz")
```
Above you say the name comes without an extension, so why not just use `.name` here to emphasize that again?

boxdot replied:
I fixed this by using the filename from the last redirect and not the filename from the source archive.

boxdot commented:
@sschuberth It looks like there are tests failing recently introduced by changes on master.

sschuberth commented:
Looks more like a temporary hiccup with Hackage, the Stack registry, or a general network issue. Locally `StackTest` passes for me, and anyway nothing goes into *master* without passing the tests first.

That said, @boxdot, please wait a bit until catching up with *master*. We need to sort out an issue first, and I'll probably get https://github.com/heremaps/oss-review-toolkit/pull/1626 merged which allows you to define "crate"-handling in `ArchiveUtils.kt only once. Afterwards I'd like to ask you a (hopefully) last time to squash all commits in this PR, rebase onto *master*, and then I believe we're good to merge.

sschuberth commented on .appveyor.yml:
```diff
@@ -13,6 +13,7 @@ environment:
   STACK_VERSION: 2.1.1
   VIRTUAL_ENV_VERSION: 15.1.0
   YARN_VERSION: 1.16.0
+  RUST_VERSION: 1.35.0
```
Nit: We generally try to keep this list sorted alphabetically for easier lookup.

sschuberth commented on .travis.yml:
```diff
@@ -33,6 +33,7 @@ env:
     - PHP_VERSION="7.1"
     - STACK_VERSION="2.1.1"
     - YARN_VERSION="1.16.0"
+    - RUST_VERSION="1.35.0"
```
Nit: We generally try to keep this list sorted alphabetically for easier lookup.

sschuberth commented on analyzer/src/main/kotlin/managers/Cargo.kt:
```diff
@@ -0,0 +1,279 @@
+
+    private fun extractVcsInfo(node: JsonNode): VcsInfo {
+        val url = extractRepositoryUrl(node)
+        val type = url?.let { VcsType.GIT } ?: VcsType.NONE // For now cargo supports only git.
```
Any particular reason why you're not using `VersionControlSystem.splitUrl()` here?

boxdot replied:
Ah, much better. Thanks.

boxdot commented:
> Looks more like a temporary hiccup with Hackage, the Stack registry, or a general network issue. Locally `StackTest` passes for me, and anyway nothing goes into _master_ without passing the tests first.
> 
> That said, @boxdot, please wait a bit until catching up with _master_. We need to sort out an issue first, and I'll probably get #1626 merged which allows you to define "crate"-handling in `ArchiveUtils.kt only once. Afterwards I'd like to ask you a (hopefully) last time to squash all commits in this PR, rebase onto _master_, and then I believe we're good to merge.

Thanks for the update. I will fix your other comments and wait for the master to catch up.

sschuberth commented:
@boxdot, from my side we're ready to again squash your PR into one commit, rebase onto master (resolving the easy conflicts in `ArchiveUtils.kt`) and merging after tests pass.

boxdot commented:
> @boxdot, from my side we're ready to again squash your PR into one commit, rebase onto master (resolving the easy conflicts in `ArchiveUtils.kt`) and merging after tests pass.

Squashed and adjusted to the changes on master. Let's hope the CI will pass.

sschuberth commented:
@boxdot, I still see the changes from the `quickcheck-state-machine` submodule in your diff, which shouldn't be. Please ensure to have run `git submodule update --init --recursive` on the state you rebase onto.

sschuberth commented:
... or simply amend your commit with the submodule being check out from master.
